common/client-core/src/client/topology_control/geo_aware_provider.rs

```diff
@@ -0,0 +1,263 @@
+            "MX" => NorthAmerica,
+            "US" => NorthAmerica,
+
+            // South America
+            "AR" => SouthAmerica,
+            "BR" => SouthAmerica,
+            "CL" => SouthAmerica,
+            "CO" => SouthAmerica,
+            "CR" => SouthAmerica,
+            "GT" => SouthAmerica,
+
+            // Oceania
+            "AU" => Oceania,
+
+            // Asia
+            "AM" => Asia,
+            "BH" => Asia,
+            "CN" => Asia,
+            "GE" => Asia,
+            "HK" => Asia,
+            "ID" => Asia,
+            "IL" => Asia,
+            "IN" => Asia,
+            "JP" => Asia,
+            "KH" => Asia,
+            "KR" => Asia,
+            "KZ" => Asia,
+            "MY" => Asia,
+            "RU" => Asia,
+            "SG" => Asia,
+            "TH" => Asia,
+            "VN" => Asia,
+
+            // Africa
+            "SC" => Africa,
+            "UG" => Africa,
+            "ZA" => Africa,
+
+            _ => {
+                info!("Unknown country code: {}", country_code);
+                Unknown
+            }
+        }
+    }
+}
+
+impl CountryGroup {
+    fn known(self) -> Option<CountryGroup> {
+        use CountryGroup::*;
+        match self {
+            Europe | NorthAmerica | SouthAmerica | Oceania | Asia | Africa => Some(self),
+            Unknown => None,
+        }
+    }
+}
+
+fn group_mixnodes_by_country_code(
+    mixnodes: Vec<PrettyDetailedMixNodeBond>,
+) -> HashMap<CountryGroup, Vec<MixId>> {
+    mixnodes
+        .into_iter()
+        .fold(HashMap::<CountryGroup, Vec<MixId>>::new(), |mut acc, m| {
+            if let Some(ref location) = m.location {
+                let country_code = location.two_letter_iso_country_code.clone();
+                if let Some(group_code) = CountryGroup::from(country_code.as_str()).known() {
+                    let mixnodes = acc.entry(group_code).or_insert_with(Vec::new);
+                    mixnodes.push(m.mix_id);
+                }
+            }
+            acc
+        })
+}
+
+fn log_mixnode_distribution(mixnodes: &HashMap<CountryGroup, Vec<MixId>>) {
+    let mixnode_distribution = mixnodes
+        .iter()
+        .map(|(k, v)| format!("{}: {}", k, v.len()))
+        .collect::<Vec<_>>()
+        .join(", ");
+    debug!("Mixnode distribution - {}", mixnode_distribution);
+}
+
+fn count_mixnodes_per_layer(mixnodes: Vec<MixNodeDetails>) -> HashMap<Layer, usize> {
+    mixnodes
+        .iter()
+        .map(|m| m.layer())
+        .fold(HashMap::<Layer, usize>::new(), |mut acc, layer| {
+            let count = acc.entry(layer).or_insert(0);
+            *count += 1;
+            acc
+        })
+}
+
+fn check_layer_integrity(mixnodes: Vec<MixNodeDetails>) -> Result<(), ()> {
+    let mut layer_counts = count_mixnodes_per_layer(mixnodes);
+    for layer in &[Layer::One, Layer::Two, Layer::Three] {
+        layer_counts.entry(*layer).or_insert(0);
+        let count = layer_counts[layer];
+        if count < 2 {
```

Comment:
why 2 : )

Reply:
I pulled it out of my behind. But let me at least extract out to a const